docs/function-math-lerp.md

```diff
@@ -0,0 +1,36 @@
+# math.lerp
+
+## Summary
+
+Add `lerp` (linear interpolation) to the standard `math` library, which takes in two numbers, as well as a fractional value that interpolates between the two.
+
+## Motivation
+
+`lerp` is a very common mathematical formula and should be natively supported in Luau for ease-of-use and faster execution. Despite potential floating-point imperfections (which we will discuss later), the Luau `math` library already contains many floating-point-based formulas and we can design `math.lerp` to work in effectively 100% of use-cases.
+
+## Design
+
+The base logic without edge-cases will look as follows, but implemented in C++. As per already-existing `:Lerp` logic in the Roblox level above Luau, this function will be **unbounded**, meaning that `alpha` does not have to strictly be a number between 0 and 1, allowing for linear extrapolation. This also means that numbers that approach `inf` or `-inf` should return `inf` or `-inf` respectively to avoid overflow.
+
+```lua
+function math.lerp(x: number, y: number, alpha: number): number
+    if x == y then return x end
+    if alpha == 0 then return x end
+    if alpha == 1 then return y end
+    
+    return x + (y - x) * alpha
+end
+```
+
+When implementing a `lerp` function, we can judge how accurate the function is by the following criteria (credit to [this](https://github.com/rust-lang/rust/issues/86269#issuecomment-869108301) post about the nature of `lerp`). If the `lerp` function can meet all of these criteria, then the function is effectively perfect.
+- Is the function **exact**? This means that `math.lerp(x, y, 0) == x` and `math.lerp(x, y, 1) == y` exactly.
+- Is the function **consistent**? This means that `math.lerp(x, x, alpha) == x`, despite the value of `alpha`.
+- Is the function **monotonic**? This means that if `y` is strictly greater than `x`, `math.lerp(x, y, alpha)` should also strictly increase as `alpha` strictly increases, and vice versa for strictly decreasing. Now, due to the limitations of `double` precision, if `x` and `y` are very close together, there may be overlap where multiple values of `alpha` may correlate to the same result. This is ultimately unavoidable with any kind of `double` and `float` math, as we can observe the same behavior with `math.sin`, `math.cos`, etc.
+
+## Drawbacks
+
+As mentioned in the [Design](##design) section, the naïve implementation of `lerp` may introduce precision error, and all of the edge-cases will need to be accounted for.
```

Comment:
```lua
-- lerp(a, b, t):
-- t can be much higher precision when near 0
-- so when t is near 0, we don't ever want to interpolate from a1 to a0
-- when t is closer to 0, we want higher accuracy around a
-- when t is closer to 1, we want higher accuracy around b
-- we want equations of the form A + t*B, as this guarantees monotonic behavior
-- at the interchange point, t = 1/2, we want to guarantee monotonic behavior
-- so we bound the result with respect to (a + b)/2

-- some things to consider:
-- (100000000 samples)

-- given x, y = sortabs(random()*randomsign(), random()*randomsign()):
--   x - y + y ~= x: ~49.2%
--   y - x + x ~= y: ~20.6%

-- given x, y = sortabs(random(), random()):
--   x - y + y ~= x: ~30.1%
--   y - x + x ~= y: ~3.98%

-- given x, y = sortabs(random()*1.5, random()*1.5):
--   x - y + y ~= x: ~25.4%
--   y - x + x ~= y: ~1.63%

-- given floorlog2(x) == floorlog2(y) == floorlog2(y - x)
--   x - y + y ~= x: 0%
--   y - x + x ~= y: 0%

local function lerp2(a0, a1, t)
	local m = (a0 + a1)/2
	if t <= 1/2 then
		local a = a0 + (a1 - a0)*t
		if a0 < a1 then
			return math.min(m, a)
		else
			return math.max(m, a)
		end
	else
		local a = a1 + (a1 - a0)*(t - 1)
		if a0 < a1 then
			return math.max(m, a)
		else
			return math.min(m, a)
		end
	end
end

-- but after testing, it appears that this never seems to expose un-monotonic behavior
-- Perhaps it has something to do with some specifics on how floating point rounds
local function lerp2(a0, a1, t)
	if t <= 1/2 then
		return a0 + (a1 - a0)*t
	else
		return a1 + (a1 - a0)*(t - 1)
	end
end
```

Reply:
so from that, seems like at *most* all we'd need is to conditionally swap `a` <-> `b` and invert `t` to reach most/all the guarantees we wanted? aka no reason to go more complex than `lerp2`?

Reply:
blending in @/vegorov-rbx's comments, would it even be worth special casing anything? would the only issue with using `a0 + (a1 - a0) * t` be that `t` has less precision around `1`?

Reply:
`t` has less precision around `1`, but there's no way to extract more precision from `t` in the first place, what they give is what we get to work with.

Reply:
From a user perspective, if I want more accuracy around `a1` than `a0`, I would need to find some accurate way of generating `s = 1 - t` directly, and then use `math.lerp(a1, a0, s)`

Reply:
Oh right this is about the precision of the multiplication with `(a1 - a0)`, right? We want to take advantage of the higher density of floating point numbers near 0 so we would rather "multiply down". e.g.: `x * 0.01` will be more precise than `x - (x * 0.99)`.

All makes sense - `lerp2` seems like the nicest "correct" solution. I suppose it just depends on whether we would want to use a conditional or whether that's deemed too heavy - will let the Luau folk speak to that.

Reply:
Oh this is actually really clever, so the only "inaccuracy" we get is near 0.5

Reply:
Thank you for this new insight, I will try and push this forward :)